lammps/lammps
# Add MPI-enabled unit tests (C/C++ only)

**Summary**

Adds first set of MPI-enabled unit tests. See `c-library/test_library_mpi.cpp`

- add `sublo`, `subhi`, `sublo_lambda`, and `subhi_lambda` to `lammps_extract_global()` and `lammps_extract_global_datatype()`
- add documentation for those flags
- add simple test for global bounding box and subboxes (orthogonal and triclinic)
- test `lammps_gather()` and `lammps_scatter()` methods
- defines a new `add_mpi_test()` CMake function
- add `universe_size`, `universe_rank`, `world_size`, and `world_rank` to `lammps_extract_setting()`
- add tests for above
- add tests for launching LAMMPS through the library interface on a split communicator and in multi-partition mode (2x2 each)
- add support for extracting additional variable styles as char pointer to `lammps_extract_variable()`
- add check for using a "universe" style variable in the multi-partition tests.
- add support for `-in none` so that multi-partition runs can be launched from the library interface without an (initial or dummy) input file.

**Author(s)**

Richard Berger (Temple University) @rbberger 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**



**Implementation Notes**

Replaces the Google Test default listeners/printer with a custom implementation `MPIPrinter`, defined in the `testing/mpitesting.h` header. Include `test_mpi_main.h` to define a MPI-enabled `main()` method that does the necessary setup.

Rank 0 collects errors from other ranks. That means if an assertion fails on any rank, it will be shown in the output once.
Error messages are will have "[Rank #]` prepended to indicate ownership of the failed assertion.

The `add_mpi_test(name binary nprocs)` CMake function allows adding a test with given `name`, running a `binary` with `nprocs` processors.

**Post Submission Checklist**



- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**
